lpc55_sign/src/signed_image.rs

```diff
@@ -24,333 +22,115 @@ fn get_pad(val: usize) -> usize {
-                .unwrap();
-
-        // We need the sha256 of the pubkeys. This is just the sha256
-        // of n + e from the pubkey
-        let n = root0_pubkey.n();
-        let e = root0_pubkey.e();
-
-        let mut sha = sha2::Sha256::new();
-        sha.update(&n.to_bytes_be());
-        sha.update(&e.to_bytes_be());
-        let result = sha.finalize();
-        let mut v: Vec<u8> = vec![0; 32];
-        v.copy_from_slice(&result);
-        v
-    });
-
-    // We're relying on packed_struct to catch errors of padding
-    // or size since we know how big this should be
-    let cert_header_size = CertHeader::packed_bytes_size(None)?;
-
-    let mut new_cert_header: CertHeader = CertHeader::new(cert_header_size, cert_len);
-
-    new_cert_header.certificate_count = signing_chain.cert_paths.len() as u32;
-
-    // some math on how many bytes we sign
-    //
-    // Base image + padding
-    // certificate header block
-    // certificate length
-    // 4 sha256 hashes
-    let signed_len = bytes.len()
-        + image_pad
-        + cert_header_size
-        + (new_cert_header.certificate_table_len as usize)
-        + 32 * 4;
-
-    // Total image length includes 256 bytes of signature
-    let total_len = signed_len + 256;
-
-    new_cert_header.total_image_len = signed_len.try_into().unwrap();
-
-    let image_len = bytes.len();
-
-    byteorder::LittleEndian::write_u32(&mut bytes[0x20..0x24], total_len as u32);
-
-    let boot_field = BootField::new(BootImageType::SignedImage);
-
-    bytes[0x24..0x28].clone_from_slice(&boot_field.pack()?);
-    // Our execution address goes in the next word
-    byteorder::LittleEndian::write_u32(&mut bytes[0x34..0x38], execution_address);
-    // where to find the block. For now just stick it right after the image
-    byteorder::LittleEndian::write_u32(&mut bytes[0x28..0x2c], (image_len + image_pad) as u32);
-
-    let mut out = OpenOptions::new()
-        .write(true)
-        .truncate(true)
-        .append(false)
-        .create(true)
-        .open(outfile_path)?;
-
-    // Generate the image, see 7.3.4 of v2.4 UM 11126 for the layout
-    out.write_all(&bytes)?;
-    if image_pad > 0 {
-        out.write_all(&vec![0; image_pad])?;
-    }
-    out.write_all(&new_cert_header.pack()?)?;
-    out.write_all(&signing_certs)?;
-    let mut rkth_sha = sha2::Sha256::new();
-    for c in root_hashes {
-        out.write_all(&c)?;
-        rkth_sha.update(&c);
+) -> Result<(Vec<u8>, [u8; 32])> {
+    if signing_certs.len() < 1 {
+        return Err(anyhow!("Need at least a root certificate"));
     }
-    for _ in 0..(4 - root_cnt) {
-        let empty_hash: [u8; 32] = [0; 32];
-        out.write_all(&empty_hash)?;
-        rkth_sha.update(empty_hash);
+    if root_certs
+        .iter()
+        .find(|&&cert| cert == signing_certs[0])
+        .is_none()
+    {
+        return Err(anyhow!("Root signing certificate must appear in root list"));
```

Comment:
I like
```suggestion
        bail!("Root signing certificate must appear in root list");
```
but your taste may vary

Reply:
Me too, now! Not a serious `anyhow` user yet, thanks for the tip. Applied in c5fe032.